runtime/js/40_http.js

```diff
@@ -0,0 +1,215 @@
+      } catch (error) {
+        if (error instanceof errors.BadResource) {
+          return { value: undefined, done: true };
+        } else if (error instanceof errors.Interrupted) {
+          return { value: undefined, done: true };
+        }
+        throw error;
+      }
+    }
+
+    return(value) {
+      core.close(this.#rid);
+      return Promise.resolve({ value, done: true });
+    }
+
+    [Symbol.asyncIterator]() {
+      return this;
+    }
+  }
+
+  function readRequest(requestRid, zeroCopyBuf) {
+    return Deno.core.jsonOpAsync(
+      "op_http_read_request",
+      requestRid,
+      zeroCopyBuf,
+    );
+  }
+
+  function respond(responseSenderRid, resp, zeroCopyBuf) {
+    return Deno.core.jsonOpSync("op_http_respond", [
+      responseSenderRid,
+      resp.status ?? 200,
+      flatEntries(resp.headers ?? {}),
+    ], zeroCopyBuf);
+  }
+
+  function writeResponse(responseBodyRid, zeroCopyBuf) {
+    return Deno.core.jsonOpAsync(
+      "op_http_write_response",
+      responseBodyRid,
+      zeroCopyBuf,
+    );
+  }
+
+  function createRespondWith(responseSenderRid, connRid) {
+    return async function (resp) {
+      if (resp instanceof Promise) {
+        resp = await resp;
+      }
+
+      if (!(resp instanceof Response)) {
+        throw new TypeError(
+          "First argument to respondWith must be a Response or a promise resolving to a Response.",
+        );
+      }
+      // If response body is Uint8Array it will be sent synchronously
+      // in a single op, in other case a "response body" resource will be
+      // created and we'll be streaming it.
+      const body = resp[fastBody]();
+      let zeroCopyBuf;
+      if (body instanceof ArrayBuffer) {
+        zeroCopyBuf = new Uint8Array(body);
+      } else if (!body) {
+        zeroCopyBuf = new Uint8Array(0);
+      } else {
+        zeroCopyBuf = null;
+      }
+
+      const responseBodyRid = respond(
+        responseSenderRid,
+        resp,
+        zeroCopyBuf,
+      );
+
+      // If `respond` returns a responseBodyRid, we should stream the body
+      // to that resource.
+      if (typeof responseBodyRid === "number") {
+        if (!body || !(body instanceof ReadableStream)) {
+          throw new Error(
+            "internal error: recieved responseBodyRid, but response has no body or is not a stream",
+          );
+        }
+        for await (const chunk of body) {
+          const data = new Uint8Array(
+            chunk.buffer,
+            chunk.byteOffset,
+            chunk.byteLength,
+          );
+          await writeResponse(
+            responseBodyRid,
+            data,
+          );
+        }
+
+        // Once all chunks are sent, and the request body is closed, we can close
+        // the response body.
+        core.close(responseBodyRid);
+        await Deno.core.jsonOpAsync("op_http_poll", connRid);
+      }
+    };
+  }
+
+  class RequestBodyStream extends ReadableStream {
```

Comment:
This is not spec compliant. The prototype chain will be wrong, and webidl won't be happy about it. Please revert to the plain `ReadableStream`. Instead of a class you can make this a helper function that constructs the `ReadableStream`.

Reply:
fixed